\section{Introduction}

HMMER is an implementation of profile hidden Markov models (profile
HMMs) for biological sequence analysis
\cite{Krogh94,Eddy98,Durbin98}. 

\subsection{How to avoid reading this manual}

I hate reading documentation. If you're like me, you're sitting here
thinking, \pageref{manualend} pages of documentation, you must be
joking! I just want to know that the software compiles, runs, and
gives apparently useful results, before I read some horrible
\pageref{manualend} pages of someone's documentation. for you bitter
cynics like me that have seen one too many software packages that
don't work:

\begin{itemize}
\item Follow the quick installation instructions on page
      \pageref{section:installation}. An automated test suite
      is included, so you will know immediately if something
      went wrong.
\item Go to the tutorial, section \ref{section:tutorial} on page
\pageref{section:tutorial}, which walks you through some examples of
using HMMER on real data.
\end{itemize}

Everything else, you can come back and read later.

\subsection{What profile HMMs are}

Profile HMMs are statistical models of multiple sequence alignments.
They capture position-specific information about how conserved each
column of the alignment is, and which residues are likely.  Anders
Krogh, David Haussler, and co-workers at UC Santa Cruz introduced
profile HMMs to computational biology \cite{Krogh94}, adopting HMM
techniques which have been used for years in speech recognition. HMMs
had been used in biology before the Krogh/Haussler work, notably by
Gary Churchill \cite{Churchill89}, but the Krogh paper had a dramatic
impact, because HMM technology was so well-suited to the popular
``profile'' methods for searching databases using multiple sequence
alignments instead of single query sequences. 

``Profiles'' were introduced by Gribskov and colleagues
\cite{Gribskov87,Gribskov90}, and several other groups introduced
similar approaches at about the same time, such as ``flexible
patterns'' \cite{Barton90}, and
``templates''\cite{Bashford87,Taylor86}. The term ``profile'' has
stuck.\footnote{There has been some agitation to call all such models
``position specific scoring matrices'', PSSMs, but certain small
nocturnal North American marsupials have a prior claim on the name.}
All of the profile methods are more or less statistical descriptions
of the consensus of a multiple sequence alignment. They use
\emph{position-specific} scores for amino acids (or nucleotides) and
position specific penalties for opening and extending an insertion or
deletion.  Traditional pairwise alignment (for example, BLAST
\cite{Altschul90}, FASTA \cite{Pearson88}, or the Smith/Waterman
algorithm \cite{Smith81}) uses position-{\em independent} scoring
parameters. This property of profiles captures important information
about the degree of conservation at various positions in the multiple
alignment, and the varying degree to which gaps and insertions are
permitted.

The advantage of using HMMs is that HMMs have a formal probabilistic
basis. We use probability theory to guide how all the scoring
parameters should be set. Though this might sound like a purely
academic issue, this probabilistic basis lets us do things that more
heuristic methods cannot do easily. For example, a profile HMM can be
trained from unaligned sequences, if a trusted alignment isn't yet
known. Another consequence is that HMMs have a consistent theory
behind gap and insertion scores. In most details, profile HMMs are a
slight improvement over a carefully constructed profile -- but less
skill and manual intervention are necessary to use profile HMMs.  This
allows us to make libraries of hundreds of profile HMMs and apply them
on a very large scale to whole genome analysis.  One such database of
protein domain models is Pfam \cite{Sonnhammer97,Bateman02}, which is
a significant part of the Interpro protein domain annotation system
\cite{Mulder03}. The construction and use of Pfam is tightly tied to
the HMMER software package.

HMMs do have important limitations. One is that HMMs do not capture
any higher-order correlations.  An HMM assumes that the identity of a
particular position is independent of the identity of all other
positions.\footnote{This is not strictly true. There is a subtle
difference between an HMM's state path (a first order Markov chain)
and the sequence described by an HMM (generated from the state path by
independent emissions of symbols at each state).} HMMs make poor
models of RNAs, for instance, because an HMM cannot describe base
pairs. Also, compare protein ``threading'' methods, which usually
include scoring terms for nearby amino acids in a three-dimensional
protein structure.

\subsection{Applications of profile HMMs}

One application of HMMER is when you are working on a protein sequence
family, and you have carefully constructed a multiple sequence
alignment. Your family, like most protein families, has a number of
strongly (but not absolutely) conserved key residues, separated by
characteristic spacing. You wonder if there are more members of your
family in the sequence databases, but the family is so evolutionarily
diverse, a BLAST search with any individual sequence doesn't even find
the rest of the sequences you already know about. You're sure there
are some distantly related sequences in the noise. You spend many
pleasant evenings scanning weak BLAST alignments by eye to find ones
with the right key residues are in the right places, but you wish
there was a computer program that did this a little better.

Another application is the automated annotation of the domain
structure of proteins. Large databases of curated alignments and HMMER
models of known domains are available, including Pfam \cite{Bateman02}
and SMART \cite{LetunicBork02} in the Interpro database consortium
\cite{Mulder03}. (Many ``top ten protein domains'' lists, a standard
table in genome analysis papers, rely heavily on HMMER annotation.)
Say you have a new sequence that, according to a BLAST analysis, shows
a slew of hits to receptor tyrosine kinases. Before you decide to call
your sequence an RTK homologue, you suspiciously recall that RTK's
are, like many proteins, composed of multiple functional domains, and
these domains are often found promiscuously in proteins with a wide
variety of functions. Is your sequence really an RTK? Or is it a novel
sequence that just happens to have a protein kinase catalytic domain
or fibronectin type III domain?

And another application is the automated construction and maintenance
of large multiple alignment databases.  It is useful to organize
sequences into evolutionarily related families. But there are
thousands of protein sequence families, some of which comprise tens of
thousands of sequences -- and the primary sequence databases continue
to double every year or two. This is a hopeless task for manual
curation; but on the other hand, manual curation is really necessary
for high-quality, biologically relevant multiple alignments. Databases
like Pfam \cite{Bateman02} are constructed by distinguishing between a
stable curated ``seed'' alignment of a small number of representative
sequences, and ``full'' alignments of all detectable homologs; HMMER
is used to make a model of the seed, search the database for homologs,
and automatically produce the full alignment by aligning every
sequence to the seed consensus.

You may find other applications as well. Using hidden Markov models to
make a linear consensus model of a bunch of related strings is a
pretty generic problem, and not just in biological sequence analysis.
HMMER has reportedly been used to model music, speech, and automobile
engine telemetry. If you use it for something particularly strange,
I'd be curious to hear about it (but I never want to see my error
messages showing up on the console of my Honda).

\subsection{How to avoid using this software (links to similar software)}

Several implementations of profile HMM methods and related
position-specific scoring matrix methods are available.  Some are
listed in the table below.

\begin{center}
\begin{tabular}{ll}
Software  &   URL \\ \hline
HMMER     & \htmladdnormallink{http://hmmer.wustl.edu/}{http://hmmer.wustl.edu/}  \\
SAM       & \htmladdnormallink{http://www.cse.ucsc.edu/research/compbio/sam.html}{http://www.cse.ucsc.edu/research/compbio/sam.html} \\
PFTOOLS   & \htmladdnormallink{http://www.isrec.isb-sib.ch/ftp-server/pftools/}{http://www.isrec.isb-sib.ch/ftp-server/pftools/}  \\
HMMpro    & \htmladdnormallink{http://www.netid.com/html/hmmpro.html}{http://www.netid.com/html/hmmpro.html}\\
GENEWISE  & \htmladdnormallink{http://www.ebi.ac.uk/Wise2/}{http://www.ebi.ac.ukWise2/} \\
PROBE     & \htmladdnormallink{ftp://ftp.ncbi.nih.gov/pub/neuwald/probe1.0/}{ftp://ftp.ncbi.nih.gov/pub/neuwald/probe1.0/} \\
META-MEME & \htmladdnormallink{http://metameme.sdsc.edu/}{http://metameme.sdsc.edu/} \\
BLOCKS    & \htmladdnormallink{http://www.blocks.fhcrc.org/}{http://www.blocks.fhcrc.org/} \\
PSI-BLAST & \htmladdnormallink{http://www.ncbi.nlm.nih.gov/BLAST/newblast.html}{http://www.ncbi.nlm.nih.gov/BLAST/newblast.html} \\
\end{tabular}
\end{center}

HMMER, SAM, PFTOOLS, and HMMpro are the most closely related to the
profile HMM methods introduced by Krogh et al. HMMpro is commercial,
not free software.

\subsection{How to learn more about profile HMMs}

Reviews of the profile HMM literature have been written by myself
\cite{Eddy96,Eddy98} and by Anders Krogh \cite{Krogh98}.

For details on how profile HMMs and probabilistic models are used in
computational biology, see the pioneering 1994 paper from Krogh et
al. \cite{Krogh94} or our book \emph{Biological Sequence Analysis:
Probabilistic Models of Proteins and Nucleic Acids} \cite{Durbin98}.

To learn more about HMMs from the perspective of the speech
recognition community, an excellent tutorial introduction has been
written by Rabiner \cite{Rabiner89}.

\begin{srefaq}{How do I cite HMMER?}
There is still no ``real'' paper on HMMER. If you're writing for an
enlightened (url-friendly) journal, the best reference is
\htmladdnormallink{http://hmmer.wustl.edu/}{http://hmmer.wustl.edu/}.
If you must use a paper reference, the best one to use is my 1998
profile HMM review \cite{Eddy98}.
\end{srefaq}
