#! /usr/bin/perl

# Test of hmmbuild/nhmmer as used to build a DNA model, then query a  
# a database of long (1MB).
#
# Usage:   ./i018-nhmmer-generic.pl <builddir> <srcdir> <tmpfile prefix>
# Example: ./i018-nhmmer-generic.pl ..         ..       tmpfoo
#
BEGIN {
    $builddir  = shift;
    $srcdir    = shift;
    $tmppfx    = shift;
    $verbose   = shift;  # if arg not given, defaults to false (zero)
}


# The test makes use of the following file:
#
# 3box.sto              <msafile>  Single 3box alignment

# It creates the following files:
# $tmppfx.hmm           <hmm>     1 model, 3box
# $tmppfx.A             <seqfile> 1 random seq, ~4.5MB in length
# $tmppfx.B             <seqfile> 2 random seqs, generated by hmmemit from $tmppfx.hmm 
# $tmppfx.fa            <seqdb>   Roughly 4.5MB of a single pseudochromosome, consisting of the two sequences from $tmppfx.B inserted into the sequence of $tmppfx.A  

# All models assumed to be in testsuite subdirectory.
$alignment   = "3box.sto";

@h3progs =  ( "hmmemit", "hmmbuild", "nhmmer");
@eslprogs =  ("esl-shuffle");

# Verify that we have all the executables and datafiles we need for the test.
foreach $h3prog  (@h3progs)  { if (! -x "$builddir/src/programs/$h3prog")         { die "FAIL: didn't find $h3prog executable in $builddir/src/programs\n";              } }
foreach $eslprog (@eslprogs) { if (! -x "$builddir/lib/easel/miniapps/$eslprog")  { die "FAIL: didn't find $eslprog executable in $builddir/lib/easel/miniapps\n";  } }

if (! -r "$srcdir/testsuite/$alignment")  { die "FAIL: can't read msa $alignment in $srcdir/testsuite\n"; }

# Create the test hmm
$cmd = "$builddir/src/programs/hmmbuild $tmppfx.hmm $srcdir/testsuite/$alignment";
$output = do_cmd($cmd);
if ($? != 0) { die "FAIL: hmmbuild failed unexpectedly\n"; } 
if ($output !~ /1     3box                    22    22    20    75    22.00  1.415/) {
	die "FAIL: hmmbuild failed to build correctly\n";
}
$output = do_cmd( "grep MAXL $tmppfx.hmm" );
if ($output !~ /MAXL  75/) {
    die "FAIL: hmmbuild failed to build correctly\n";
}


# Create a roughly 4.5MB database against which to search
$database   = "$tmppfx.fa";
do_cmd ( "$builddir/lib/easel/miniapps/esl-shuffle --seed 33 --dna -G -N 1 -L 4500000 -o $tmppfx.A" );
do_cmd ( "$builddir/src/programs/hmmemit -N 2 --seed 4 $tmppfx.hmm >  $tmppfx.B " );
do_cmd ( "$builddir/src/programs/hmmemit -N 1 --seed 3 $tmppfx.hmm >> $tmppfx.B" ); 
do_cmd ( "head -n 33000 $tmppfx.A > $database" );
do_cmd ( "head -n 2 $tmppfx.B | tail -n 1 >> $database" );
do_cmd ( "tail -n +33001 $tmppfx.A | head -n 22000 >> $database");
do_cmd ( "head -n 4 $tmppfx.B | tail -n 1 >> $database" );
do_cmd ( "tail -n 20000 $tmppfx.A >> $database" );
do_cmd ( "tail -n 1 $tmppfx.B >> $database" );

# perform nhmmer search
$cmd = "$builddir/src/programs/nhmmer --tformat fasta $tmppfx.hmm $database";
$output = do_cmd($cmd);

if ($? != 0) { die "FAIL: nhmmer failed unexpectedly\n"; }
$expect = 
q[Target sequences:                          1  \(9000000 residues searched\)
Residues passing MSV filter:            174223  \(0.0194\); expected \(0.02\)
Residues passing bias filter:           145398  \(0.0162\); expected \(0.02\)
Residues passing Vit filter:              6173  \(0.000686\); expected \(0.001\)
Residues passing Fwd filter:               276  \(3.07e-05\); expected \(1e-05\)
Total number of hits:                        4  \(8.78e-06\)];
if ($output !~ /$expect/s) {
    die "FAIL: nhmmer failed search test 1\nExpect:\n$expect\nObserved:\n$output\n";
}

$expect = 
  q[  0.031   17.2   0.0  random   3299961 3299979\s*
      0.046   16.6   0.0  random   1979941 1979960\s*
      0.066   16.1   0.0  random   4499980 4500000\s*
       0.23   14.5   0.0  random   1518622 1518604\s*];          
        
        
if ($output !~ /$expect/s) {
    die "FAIL: nhmmer failed search test 2\nExpect:\n$expect\nObserved:\n$output\n";
}

$cmd = "$builddir/src/programs/nhmmer --tformat fasta --toponly $tmppfx.hmm $database";
$output = do_cmd($cmd);
if ($? != 0) { die "FAIL: nhmmer failed unexpectedly\n"; }
$expect = 
q[Target sequences:                          1  \(4500000 residues searched\)
Residues passing MSV filter:             84348  \(0.0187\); expected \(0.02\)
Residues passing bias filter:            70876  \(0.0158\); expected \(0.02\)
Residues passing Vit filter:              3091  \(0.000687\); expected \(0.001\)
Residues passing Fwd filter:               155  \(3.44e-05\); expected \(1e-05\)
Total number of hits:                        3  \(1.33e-05\)];


if ($output !~ /$expect/s) {
    die "FAIL: nhmmer failed search test 3\nExpect:\n$expect\nObserved:\n$output\n";
}
$expect = 
  q[  0.015   17.2   0.0  random   3299961 3299979\s*
      0.023   16.6   0.0  random   1979941 1979960\s*
      0.033   16.1   0.0  random   4499980 4500000]; 
if ($output !~ /$expect/s) {
    die "FAIL: nhmmer failed search test 4\nExpect:\n$expect\nObserved:\n$output\n";
}

print "ok.\n";

if (!$verbose) {
  unlink "$tmppfx.hmm";
  unlink "$tmppfx.A";
  unlink "$tmppfx.B";
  unlink "$tmppfx.fa";
}
exit 0;


sub do_cmd {
    $cmd = shift;
    print "$cmd\n" if $verbose;
    return `$cmd`;	
}
